nexus/types/src/external_api/params.rs

```diff
@@ -18,34 +19,109 @@ use serde::{
 use std::{net::IpAddr, str::FromStr};
 use uuid::Uuid;
 
+#[derive(Deserialize, JsonSchema)]
+pub struct OrganizationPath {
+    pub organization: NameOrId,
+}
+
+#[derive(Deserialize, JsonSchema)]
+pub struct ProjectPath {
+    pub project: NameOrId,
+}
+
+#[derive(Deserialize, JsonSchema)]
+pub struct InstancePath {
+    pub instance: NameOrId,
+}
+
+#[derive(Deserialize, JsonSchema)]
+pub struct OrganizationSelector {
+    pub organization: NameOrId,
+}
+
+impl From<Name> for OrganizationSelector {
+    fn from(name: Name) -> Self {
+        OrganizationSelector { organization: name.into() }
+    }
+}
+
+#[derive(Deserialize, JsonSchema)]
+pub struct OptionalOrganizationSelector {
+    #[serde(flatten)]
+    pub organization_selector: Option<OrganizationSelector>,
+}
+
 #[derive(Deserialize, JsonSchema)]
 pub struct ProjectSelector {
+    #[serde(flatten)]
+    pub organization_selector: Option<OrganizationSelector>,
     pub project: NameOrId,
-    pub organization: Option<NameOrId>,
+}
+
+// TODO-v1: delete this post migration
+impl ProjectSelector {
+    pub fn new(organization: Option<NameOrId>, project: NameOrId) -> Self {
+        ProjectSelector {
+            organization_selector: organization
+                .map(|o| OrganizationSelector { organization: o }),
+            project,
+        }
+    }
+}
+
+#[derive(Deserialize, JsonSchema)]
+pub struct ProjectList {
+    #[serde(flatten)]
+    pub pagination: PaginatedByNameOrId,
+    #[serde(flatten)]
+    pub organization: OrganizationSelector,
+}
```

Comment:
I'd almost rather define this next to the endpoint handler since that's the only place it's called.

Reply:
Yeah, I just grouped them together because I'm hoping to macroize it at some point.

Reply:
@david-crespo I thought you were the one who split these out by type rather than by use?